src/Functions/FunctionsPunycode.cpp

```diff
@@ -0,0 +1,329 @@
+                               (0x80 | ((u32 >> 6) & 0x3F)),(0x80 | (u32 & 0x3F)));
+            }
+        }
+        return pos;
+    }
+
+    size_t punycodeEncodeInternal(const std::u32string & input,
+                                  UInt8 * encoded_string) noexcept
+    {
+        char buf[labelMaxLength]{};
+        size_t prev{0}, length{0}, encoded_length {0};
+        for (size_t size = input.size(), idx = 0; idx < size; ++idx)
+        {
+            if ('.' == input[idx])
+            {
+                if (!isASCIIStrUTF32(input, prev, idx))
+                {
+                    const auto result_code = punycode_encode(idx - prev, input.data() + prev,
+                                                             nullptr, &(length = labelMaxLength), buf);
+                    if (punycode_success != result_code)
+                        return 0;
+
+                    std::copy_n(acePrefix.data(), acePrefix.size(), encoded_string + encoded_length);
+                    encoded_length += acePrefix.size();
+                    std::copy_n(buf, length, encoded_string + encoded_length);
+                    encoded_length += length;
+                }
+                else
+                {
+                    encoded_length += appendUTF32ToString(input.data(), prev, idx, encoded_string + encoded_length);
+                }
+
+                encoded_string[encoded_length++] = '.';
+                // To skip '.' with prefix increment '++idx'
+                prev = ++idx;
+            }
+        }
+
+        if (!isASCIIStrUTF32(input, prev, input.size()))
+        {
+            const auto result_code = punycode_encode(input.length() - prev, input.data() + prev,
+                                                     nullptr, &(length = labelMaxLength), buf);
+            if (punycode_success != result_code)
+                return 0;
+
+            std::copy_n(acePrefix.data(), acePrefix.size(), encoded_string + encoded_length);
+            encoded_length += acePrefix.size();
+            std::copy_n(buf, length, encoded_string + encoded_length);
+            encoded_length += length;
+        }
+        else
+        {
+            encoded_length += appendUTF32ToString(input.data(), prev, input.size(), encoded_string + encoded_length);
+        }
+        return encoded_length;
+    }
+
+    /** If the string contains ' :// ' then this string contains the full DNS name
+     *  if the string do not contains ' :// ' then we consider the entire input
+     *  as data to be encoded regardless of the content of characters in it ' / '
+     */
+    constexpr std::pair<size_t, size_t> getFqdnBounds(std::string_view input) noexcept{
+        std::string::size_type start = input.find("://");
+        if (std::string::npos == start)
+            return {0, input.size() };
+        std::string::size_type last = input.find('/', start + 3);
+        return {start + 3, std::string::npos == last ? input.size() : last};
```

Comment:
This function is related to URL processing and not punycode or idna encoding. So I believe punycode/idna mustn't try to extract a domain and suppose instead that the domain is already extracted. Anyway we already have the function `domain()` in ClickHouse which can be easily combined with our function:

```
SELECT idnaDecode(domain(...))
```